### PR TITLE
update for v3

### PR DESCRIPTION
**Pull request by alex-Arc**
This PR brings the module mostly up to date with ontime v3
- [x] handle the custom fields
- [x] feedback
- [x] v3 variables
- [x] move all state to parallel structure
- [x] Presets for Playback
- [x] Presets for  Timer Management
- [x] implement #46 

### TODO
  - [ ] what should we pull in from [Runtime](https://github.com/cpvalente/ontime/blob/v3/packages/types/src/definitions/runtime/Runtime.type.ts)
  - [ ] look into companion expressions as inputs
  - [ ] Presets for Message Control
  - [ ] Presets for Aux Timer
  - [ ] Feedback from aux timer?

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -52,7 +55,6 @@ export function GetConfigFields(): SomeCompanionConfigField[] {
 			type: 'checkbox',
 			default: true,
 			width: 4,
-			isVisible: (config) => config.version === 'v2',
 			tooltip: 'Chose if you want Companion to refetch the events from Ontime when the rundown gets updated.',
```
Comment (cpvalente) [inline src/config.ts:58]:
```suggestion
			tooltip: 'Whether Companion should keep the rundown updated with Ontime by refetching on change.',
```

File: src/config.ts [hunk 2]
```diff
@@ -42,7 +42,10 @@ export function GetConfigFields(): SomeCompanionConfigField[] {
 			id: 'version',
 			type: 'dropdown',
 			default: 'v2',
-			choices: [{ id: 'v2', label: 'Ontime V2' }],
+			choices: [
+				{ id: 'v2', label: 'Ontime V2' },
+				{ id: 'v3', label: 'Ontime V3' },
+			],
 			width: 12,
 			tooltip: 'Choose wich version of Ontime you are connecting to.',
```
Comment (cpvalente) [inline src/config.ts:50]:
```suggestion
			tooltip: 'Which version of Ontime you are connecting to.',
```

File: src/v3/actions.ts [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+
+enum ActionCommand {
+	Start = 'start',
+	Load = 'load',
+	Reload = 'reload',
+	Pause = 'pause',
+	Stop = 'stop',
+	Roll = 'roll',
+	Add = 'addtime',
+	SetOnAir = 'set-onair',
```
Comment (cpvalente) [inline src/v3/actions.ts:14]: this is no longer actionable by the user

File: src/v3/actions.ts [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+			callback: () => {
+				socketSendJson(ActionCommand.Stop)
+			},
+		},
+		[ActionId.Reload]: {
+			name: 'Reload selected event',
+			options: [],
+			callback: () => {
+				socketSendJson(ActionCommand.Reload)
+			},
```
Comment (cpvalente) [inline src/v3/actions.ts:217]: suggestion: we could use the implicit return to simplify the logic and improve readability (although arguable)

```ts
callback: () => socketSendJson(ActionCommand.Reload),
```

File: src/v3/actions.ts [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+					step: 1,
+					required: true,
+				},
+			],
+			callback: (action) => {
+				let val =
+					(Number(action.options.hours) * 60 + Number(action.options.minutes)) * 60 + Number(action.options.seconds)
+				if (action.options.addremove === 'remove') {
+					val = val * -1
+				}
```
Comment (cpvalente) [inline src/v3/actions.ts:276]: suggestion, should logic be extracted away from config code?

File: src/v3/connection.ts [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+	}
+
+	const updateTimer = (val: TimerState) => {
+		const timer = val.current === null ? defaultTimerObject : toReadableTime(val.current)
+		const timer_start = val.startedAt === null ? defaultTimerObject : toReadableTime(val.startedAt)
+		const timer_finish = val.expectedFinish === null ? defaultTimerObject : toReadableTime(val.expectedFinish)
+		const added = mstoTime(val.addedTime)
+		self.setVariableValues({
+			[variableId.TimerTotalMs]: val.current ?? 0,
+			[variableId.Time]: timer.hours + ':' + timer.minutes + ':' + timer.seconds,
```
Comment (cpvalente) [inline src/v3/connection.ts:80]: suggestion: would template strings help here?
```ts
[variableId.Time]: `${timer.hours}:${timer.minutes}:${timer.seconds}`,
```

we could also: 
a) make a utility function that joins given parameters with the semicolons
```ts
const joinTime = (...args) => args.join(":")
```
b) could we destructure `timer` at the top?
c) should we have to consider padding the timer values?

File: src/v3/connection.ts [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+		self.checkFeedbacks(
+			feedbackId.MessageVisible,
+			feedbackId.ThisMessageVisible,
+			feedbackId.TimerBlackout,
+			feedbackId.TimerBlink
+		)
+	}
+
+	const updateRuntime = (_val: Runtime) => {
+		//TODO:
```
Comment (cpvalente) [inline src/v3/connection.ts:125]: question: what is updateRuntime?

File: src/v3/connection.ts [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+	ws.onmessage = (event: any) => {
+		try {
+			const data = JSON.parse(event.data)
+			const { type, payload } = data
+
+			if (!type) {
+				return
+			}
+
+			switch (type) {
```
Comment (cpvalente) [inline src/v3/connection.ts:163]: nice!

Can I suggest adding a link to the docs to make it easier to maintain?
https://docs.getontime.no/api/runtime-data/

File: src/v3/connection.ts [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+	self.log('debug', 'fetching events from ontime')
+	try {
+		const result = await axios.get(`http://${self.config.host}:${self.config.port}/data/rundown`, {
+			responseType: 'json',
+		})
+		//TODO: maybe chack if the order has actualy changed
+		// if (self.revision >= result.data.revision) {
+		// 	self.log('debug', `skipping reload no new revision ${result.data.revision}`)
+		// 	return
+		// }
```
Comment (cpvalente) [inline src/v3/connection.ts:288]: yes, I think this can be a good idea, with the same revision, the underlying dataset hasnt changed and we should avoid adding unnecessary work on the companion app

File: src/v3/feedback.ts [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+		},
+		options: [],
+		callback: (_feedback) => {
+			return self.getVariableValue(variableId.TimerAdded) != '00:00:00'
+		},
+	}
+	feedbacks[feedbackId.OnAir] = {
+		type: 'boolean',
+		name: 'Change color from onAir',
+		description: 'Change the colors of a button if onAir is turned on',
```
Comment (cpvalente) [inline src/v3/feedback.ts:65]: suggestion: on air is a piece of state which will be true if the app timer is running. how do we communicate this?

File: src/v3/state.ts [hunk 11]
```diff
@@ -0,0 +1,199 @@
+export type RuntimeStore = {
```
Comment (cpvalente) [inline src/v3/state.ts:1]: suggestion: maybe a link to the code can help maintaining?

File: src/v3/state.ts [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+	linkStart: string | null // ID of event to link to
+	timeStrategy: TimeStrategy
+	timeStart: number
+	timeEnd: number
+	duration: number
+	isPublic: boolean
+	skip: boolean
+	colour: string
+	revision: number
+	delay?: number // calculated at runtime
```
Comment (cpvalente) [inline src/v3/state.ts:124]: suggestion: from a consumer perspective, `delay` is never nullable

**Review by cpvalente — COMMENTED** [review]
Good work here

I wonder if we should be adding some documentation links to help in future
We can also try to simplify by pushing logic outside the config files into a few utilities

Remember we dont need to get the entire of Ontime working at once, as long as we dont break v2, we could bring the v3 API slowly into companion. we can also have several PRs to a release

Good stuff ⭐

File: src/v3/state.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+	eventNow: OntimeEvent | null
+	publicEventNow: OntimeEvent | null
+	eventNext: OntimeEvent | null
+	publicEventNext: OntimeEvent | null
+
+	// extra timers
+	timer1: SimpleTimerState
+}
+
+export type TimerState = {
```
Comment (cpvalente) [inline src/v3/state.ts:21]: Suggestion: maybe move all types to a separate file ontimeTypes.ts to make it easier to find and maintain

File: src/common/ontime-types.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+export enum SupportedEvent {
+	Event = 'event',
+	Delay = 'delay',
+	Block = 'block',
+}
+
+export type OntimeBaseEvent = {
+	type: SupportedEvent
+	id: string
+	after?: string // used when creating an event to indicate its position in rundown
```
Comment (cpvalente) [inline src/common/ontime-types.ts:29]: can companion create events? otherwise this is not necessary

File: src/v2/connection.ts [hunk 15]
```diff
@@ -197,21 +188,17 @@ export function socketSendChange(type: string, eventId: string, property: InputV
...
-		self.log('debug', `fetched ${result.data.length} events`)
-		self.events = []
-		self.events = result.data
-			.filter((event: OntimeEvent) => event.type === 'event')
-			.map((event: OntimeEvent) => ({
-				id: event.id,
-				label: event.cue + ' | ' + event.title,
-			}))
+		const response = await fetch(`http://${self.config.host}:${self.config.port}/events`)
+		const data = (await response.json()) as OntimeEvent[]
```
Comment (cpvalente) [inline src/v2/connection.ts:195]: when moving from axios to fetch, you will need to handle the errors yourself. 
ie: axios throws an error if the response has an error code, while fetch doesnt
https://kentcdodds.com/blog/replace-axios-with-a-simple-custom-fetch-wrapper

Comment (alex-Arc) [reply src/v2/connection.ts:195]: ah thanks
I can just check for a 200 right?

File: .eslintrc.cjs [hunk 16]
```diff
@@ -1,18 +1,19 @@
...
-				'prettier/prettier': ['warn', { endOfLine: 'auto' }],
+				'simple-import-sort/imports': 'error',
+				// 			'no-use-before-define': 'off',
+				// 			'@typescript-eslint/no-use-before-define': 'off',
+				// 			'no-unused-vars': 'off',
+				// 			'@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
+				// 			'@typescript-eslint/camelcase': 'off',
+				// 			'@typescript-eslint/ban-ts-ignore': 'off',
+				// 			'@typescript-eslint/explicit-function-return-type': 'off',
+				// 			'@typescript-eslint/no-explicit-any': 'off',
```
Comment (cpvalente) [inline .eslintrc.cjs:16]: Remember that this PR is already pretty large and we should avoid unrelated or unnecessary changes

We risk introducing bugs to little gain. It is generally best to try to keep the PRs small and the work focused

File: src/v2/connection.ts (hunk 15 above)
Comment (cpvalente) [reply src/v2/connection.ts:202]: I believe so, remember that the `ok` property already means "any 200 return"

File: .eslintrc.cjs (hunk 16 above)
Comment (alex-Arc) [reply .eslintrc.cjs:16]: hmm yeah maybe I got carried away, I will see if i can pull out the linting template

File: src/v3/feedbacks/message.ts [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+			options: [
+				{
+					type: 'dropdown',
+					id: 'source',
+					label: 'Source',
+					default: 'timer',
+					choices: [
+						{ id: 'timer', label: 'Timer' },
+						{ id: 'lower', label: 'Lower Third' },
+						{ id: 'public', label: 'Public' },
```
Comment (cpvalente) [inline src/v3/feedbacks/message.ts:31]: reminder: this has been removed

File: src/v3/feedbacks/playback.ts [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+			return ontime.state.timer.addedTime == 0
+		}
+		return false
+	}
+
+	return {
+		[feedbackId.ColorPlayback]: {
+			type: 'boolean',
+			name: 'Playback state',
+			description: 'TODO: description',
```
Comment (cpvalente) [inline src/v3/feedbacks/playback.ts:29]:
```suggestion
			description: 'Associate a colour to Ontimes playback',
```

File: src/v3/feedbacks/playback.ts [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+					],
+					default: Playback.Play,
+				},
+			],
+			callback: (feedback) => ontime.state.timer.playback === feedback.options.state,
+		},
+		[feedbackId.ColorNegative]: {
+			type: 'boolean',
+			name: 'Timer negative',
+			description: 'TODO: better wording - Change the colors of a button if the timer runs into negative time',
```
Comment (cpvalente) [inline src/v3/feedbacks/playback.ts:54]:
```suggestion
			description: 'Colour when runtime is in overtime',
```

File: src/v3/feedbacks/playback.ts [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+				bgcolor: combineRgb(255, 0, 0),
+			},
+			options: [],
+			callback: () => (ontime.state.timer.current ? ontime.state.timer.current < 0 : false),
+		},
+		[feedbackId.ColorAddRemove]: {
+			type: 'boolean',
+			name: 'Added/removed time',
+			description:
+				'TODO: better names in dropdown - Change the colors of a button according if time has been added/removed',
```
Comment (cpvalente) [inline src/v3/feedbacks/playback.ts:66]:
```suggestion
				'Indicator colour for whether timer has user added time',
```

File: src/v3/feedbacks/playback.ts [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+					],
+					default: 'both',
+				},
+			],
+			callback: addTime,
+		},
+		[feedbackId.OnAir]: {
+			type: 'boolean',
+			name: 'Change color from onAir',
+			description: 'TODO: ',
```
Comment (cpvalente) [inline src/v3/feedbacks/playback.ts:90]:
```suggestion
			description: 'Indicator colour for whether Ontime playback is running ',
```

File: src/v3/feedbacks/timerZone.ts [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+import { feedbackId } from '../../enums.js'
+
+export function createTimerZoneFeedback(ontime: { state: { companionSpecific: { timerZone: TimerZone } } }): {
+	[id: string]: CompanionFeedbackDefinition
+} {
+	return {
+		[feedbackId.TimerZone]: {
+			type: 'boolean',
+			name: 'Timer zone',
+			description: 'TODO: description',
```
Comment (cpvalente) [inline src/v3/feedbacks/timerZone.ts:12]:
```suggestion
			description: 'Timer zones use Ontimes warn and danger times to change colour depending on timer progress',
```

File: src/v3/ontime-types.ts [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+	Event = 'event',
+	Delay = 'delay',
+	Block = 'block',
+}
+
+export type OntimeBaseEvent = {
+	type: SupportedEvent
+	id: string
+	after?: string // used when creating an event to indicate its position in rundown
+}
```
Comment (cpvalente) [inline src/v3/ontime-types.ts:38]:
```suggestion
/**
 * {@link https://github.com/cpvalente/ontime/blob/8f249b9d515fb0d799514d3a67de6713f5029faf/packages/types/src/definitions/core/OntimeEvent.type.ts GitHub}.
 */
export type OntimeBaseEvent = {
	type: SupportedEvent
	id: string
	after?: string // used when creating an event to indicate its position in rundown
}
```

File: src/v3/ontime-types.ts [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+	linkStart: string | null // ID of event to link to
+	timeStrategy: TimeStrategy
+	timeStart: number
+	timeEnd: number
+	duration: number
+	isPublic: boolean
+	skip: boolean
+	colour: string
+	revision: number
+	delay?: number // calculated at runtime
```
Comment (cpvalente) [inline src/v3/ontime-types.ts:116]:
```suggestion
	delay: number // calculated at runtime
```

File: src/v3/state.ts [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+		finishedAt: null,
+		secondaryTimer: null,
+		duration: null,
+		playback: Playback.Stop,
+	},
+	onAir: false,
+	message: {
+		timer: { text: '', visible: false, blink: false, blackout: false },
+		public: { text: '', visible: false },
+		lower: { text: '', visible: false },
```
Comment (cpvalente) [inline src/v3/state.ts:21]: reminder: remove

File: src/v3/state.ts [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+		numEvents: 0,
+		selectedEventIndex: 0,
+		offset: 0,
+		plannedStart: 0,
+		actualStart: 0,
+		plannedEnd: 0,
+		expectedEnd: 0,
+	},
+	eventNow: null,
+	publicEventNow: null,
```
Comment (cpvalente) [inline src/v3/state.ts:34]: suggestion: maybe remove `publicEvent` fields?

File: src/v3/variables.ts [hunk 27]
```diff
@@ -3,25 +3,28 @@ import { variableId } from '../enums'
...
+		//clock
 		{
 			name: 'Clock (hh:mm:ss)',
 			variableId: variableId.Clock,
 		},
+		//timer.addedTime
 		{
-			name: 'Start of event (hh:mm:ss)',
-			variableId: variableId.TimerStart,
+			name: 'Time added to current event (hh:mm:ss)',
```
Comment (cpvalente) [inline src/v3/variables.ts:13]:
```suggestion
			name: 'User added time to current event (hh:mm:ss)',
```

File: src/v3/variables.ts [hunk 28]
```diff
@@ -3,25 +3,28 @@ import { variableId } from '../enums'
...
+			variableId: variableId.TimerAdded,
 		},
 		{
-			name: 'Expected finish of event (hh:mm:ss)',
-			variableId: variableId.TimerFinish,
+			name: 'Time added in its smallest form ',
```
Comment (cpvalente) [inline src/v3/variables.ts:17]:
```suggestion
Suggested change
			name: 'User added time to current event (smallest unit)',
```

File: src/v3/variables.ts [hunk 29]
```diff
@@ -3,25 +3,28 @@ import { variableId } from '../enums'
...
+			variableId: variableId.TimerAddedNice,
 		},
+		//timer.current
 		{
-			name: 'Time added to current event',
-			variableId: variableId.TimerAdded,
+			name: 'Current event timer zone (Normal/Warning/Danger/Overtime)',
```
Comment (cpvalente) [inline src/v3/variables.ts:22]:
```suggestion
			name: 'Current event timer progress (Normal/Warning/Danger/Overtime)',
```

File: src/v3/variables.ts [hunk 30]
```diff
@@ -3,25 +3,28 @@ import { variableId } from '../enums'
...
+			variableId: variableId.TimerZone,
+		},
+		{
+			name: 'Current time of event in milliseconds',
```
Comment (cpvalente) [inline src/v3/variables.ts:26]:
```suggestion
			name: 'Current timer (milliseconds)',
```

File: src/v3/variables.ts [hunk 31]
```diff
@@ -3,25 +3,28 @@ import { variableId } from '../enums'
...
+			variableId: variableId.TimerTotalMs,
 		},
 		{
 			name: 'Current time of event (hh:mm:ss)',
```
Comment (cpvalente) [inline src/v3/variables.ts:30]:
```suggestion
			name: 'Current timer (hh:mm:ss)',
```

File: src/v3/variables.ts [hunk 32]
```diff
@@ -43,82 +46,142 @@ export function variables(): CompanionVariableDefinition[] {
...
-			variableId: variableId.SpeakerNext,
+			name: 'Expected finish of event (hh:mm:ss)',
+			variableId: variableId.TimerFinish,
 		},
+		//timer.finishedAt
+		//timer.playback
 		{
-			name: 'Note of next event',
-			variableId: variableId.NoteNext,
+			name: 'State of event (Running, Paused, Stopped, Roll)',
```
Comment (cpvalente) [inline src/v3/variables.ts:59]:
```suggestion
			name: 'Playback state (Running, Paused, Stopped, Roll)',
```

File: src/v3/variables.ts [hunk 33]
```diff
@@ -43,82 +46,142 @@ export function variables(): CompanionVariableDefinition[] {
...
+			variableId: variableId.PlayState,
 		},
+		//timer.secondaryTimer
+		//timer.startedAt
 		{
-			name: 'Cue of next event',
-			variableId: variableId.CueNext,
+			name: 'Start of event (hh:mm:ss)',
```
Comment (cpvalente) [inline src/v3/variables.ts:65]:
```suggestion
			name: 'Start time of current timer (hh:mm:ss)',
```

File: src/v3/variables.ts [hunk 34]
```diff
@@ -43,82 +46,142 @@ export function variables(): CompanionVariableDefinition[] {
...
+		//message.public.text
 		{
 			name: 'Public Message',
 			variableId: variableId.PublicMessage,
 		},
+		//message.lower.text
 		{
 			name: 'Lower thirds Message',
 			variableId: variableId.LowerMessage,
 		},
```
Comment (cpvalente) [inline src/v3/variables.ts:87]: cleanup: remove public and lower

File: src/v3/variables.ts [hunk 35]
```diff
@@ -43,82 +46,142 @@ export function variables(): CompanionVariableDefinition[] {
...
+		//message.public.visible
 		{
 			name: 'Public Message Visible',
 			variableId: variableId.PublicMessageVisible,
 		},
+		//message.lower.visible
 		{
 			name: 'Lower thirds Message Visible',
 			variableId: variableId.LowerMessageVisible,
 		},
```
Comment (cpvalente) [inline src/v3/variables.ts:102]: cleanup: remove public and lower

File: src/v3/variables.ts [hunk 36]
```diff
@@ -43,82 +46,142 @@ export function variables(): CompanionVariableDefinition[] {
...
+			variableId: variableId.NoteNext,
+		},
+		//eventNext.cue
+		{
+			name: 'Cue of next event',
+			variableId: variableId.CueNext,
+		},
+		//aux timer
+		{
+			name: 'Aux timer 1 duration in milliseconds',
```
Comment (cpvalente) [inline src/v3/variables.ts:162]:
```suggestion
			name: 'Aux timer 1 duration (milliseconds)',
```

File: src/v3/variables.ts [hunk 37]
```diff
@@ -43,82 +46,142 @@ export function variables(): CompanionVariableDefinition[] {
...
+			variableId: variableId.AuxTimerDurationMs + '-1',
+		},
+		{
+			name: 'Aux timer 1 current in milliseconds',
```
Comment (cpvalente) [inline src/v3/variables.ts:166]:
```suggestion
			name: 'Aux timer 1 current (milliseconds)',
```

File: src/v3/variables.ts [hunk 38]
```diff
@@ -43,82 +46,142 @@ export function variables(): CompanionVariableDefinition[] {
...
+		{
+			name: 'Aux timer 1 current (hh:mm:ss)',
+			variableId: variableId.AuxTimerCurrent + '-1',
+		},
+		{
+			name: 'Aux timer 1 playback',
+			variableId: variableId.AuxTimerPalyback + '-1',
+		},
+		{
+			name: 'Aux timer 1 direction',
```
Comment (cpvalente) [inline src/v3/variables.ts:182]:
```suggestion
			name: 'Aux timer 1 direction (count-up count-down)',
```

File: .github/workflows/build.yml [hunk 39]
```diff
@@ -0,0 +1,29 @@
+name: build
+
+on: [push]
```
Comment (cpvalente) [inline .github/workflows/build.yml:3]: https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow

File: src/v3/ontimev3.ts [hunk 40]
```diff
@@ -6,23 +6,27 @@ import {
...
+import { CustomFields, OntimeEvent } from './ontime-types'
 import { stateobj } from './state'
 
-export class OntimeV2 implements OntimeClient {
+import { actions } from './actions/index'
+import { feedbacks } from './feedbacks/index'
+import { variables } from './variables'
+import { presets } from './presets'
+
+export class OntimeV3 implements OntimeClient {
```
[the remark this answers is not in the document]
Comment (alex-Arc) [reply src/v3/ontimev3.ts:18]: save this for next PR

File: src/v3/connection.ts [hunk 41]
```diff
@@ -0,0 +1,383 @@
...
+			[variableId.TimerBlink]: val.timer.blink,
+		})
+
+		self.checkFeedbacks(feedbackId.MessageVisible, feedbackId.TimerBlackout, feedbackId.TimerBlink)
+	}
+
+	const updateRuntime = (val: Runtime) => {
+		ontime.state.runtime = val
+
+		//TODO:
```
[the remark this answers is not in the document]
Comment (alex-Arc) [reply src/v3/connection.ts:126]: save this for next PR

**Comment by alex-Arc** [comment]
I think we are at a good point now

File: companion/HELP.md [hunk 42]
```diff
@@ -5,109 +5,7 @@ its [Websocket API](https://cpvalente.gitbook.io/ontime/control-and-feedback/web
 
 ## Configuration
 
-- This module requires Ontime V 1 or Ontime V 2.8.0 or above
+- This module version requires Ontime v3 or above
 - You can download ontime at [www.getontime.no](https://www.getontime.no/)
 - Ontime must be reachable through the network by the device running companion
-- **Ontime server IP** should be set to the IP address of the machine running ontime, available interfaces are
-  shown in the [Info Panel](https://cpvalente.gitbook.io/ontime/main-concepts/interface-1/info)
 - **Ontime server port** is usually 4001, with the exceptions of custom and docker setups
```
Comment (cpvalente) [inline companion/HELP.md:11]:
```suggestion
- **Ontime server port** which is by default 4001. Keep in mind that this can be changed by the user
```

File: package.json [hunk 43]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "getontime-ontime",
-	"version": "3.1.3",
+	"version": "3.3.0-beta.1",
```
Comment (cpvalente) [inline package.json:3]: question: should this not be the same as the manifest.json? (3.2.0)

File: src/utilities.ts [hunk 44]
```diff
@@ -1,76 +1,86 @@
...
+	if (timer > currentEvent.timeWarning) {
+		return TimerZone.Normal
+	}
+	if (timer > currentEvent.timeDanger) {
+		return TimerZone.Warning
+	}
+	if (timer > 0) {
+		return TimerZone.Danger
+	}
+	return TimerZone.Overtime
```
Comment (cpvalente) [inline src/utilities.ts:85]: note: I am unconvinced that this is always correct. However it seems that it will be mostly correct.
Lets go with this now and keep an eye for a future improvement

**Review by cpvalente — APPROVED** [review]
Good work @alex-Arc

File: package.json (hunk 43 above)
Comment (alex-Arc) [reply package.json:3]: yes you are correnct

File: src/utilities.ts (hunk 44 above)
Comment (alex-Arc) [reply src/utilities.ts:85]: Fair :)
